server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -580,20 +597,93 @@ public synchronized void close() throws IOException {
+            new Thread(() -> {
+                try {
+                    stopperEntry.getValue().run();
+                } catch (Exception ex) {
+                    logger.warn("unexpected exception in shutdown task [" + stopperEntry.getKey() + "]", ex);
+                } finally {
+                    future.complete(null);
+                }
+            }, stopperEntry.getKey()).start();
+            futures.put(stopperEntry.getKey(), future);
+        }
+
+        @SuppressWarnings(value = "rawtypes") // Can't make an array of parameterized types, but it complains if you leave the type out
+        CompletableFuture<Void> allStoppers = CompletableFuture.allOf(futures.values().toArray(new CompletableFuture[stoppers.size()]));
+
+        TimeValue maxTimeout = MAXIMUM_SHUTDOWN_TIMEOUT_SETTING.get(this.settings());
         try {
-            stopper.get();
-        } catch (Exception e) {
-            logger.warn("unexpected exception while waiting for http server to close", e);
+            if (maxTimeout.millis() == 0) {
+                FutureUtils.get(allStoppers);
+            } else {
+                FutureUtils.get(allStoppers, maxTimeout.millis(), TimeUnit.MILLISECONDS);
+            }
+
+        } catch (ElasticsearchTimeoutException t) {
+            var unfinishedTasks = futures.entrySet()
+                .stream()
+                .filter(entry -> entry.getValue().isDone() == false)
+                .map(Map.Entry::getKey)
+                .toList();
+            logger.info("timed out while waiting for graceful shutdown tasks: " + unfinishedTasks);
+        }
+    }
+
+    private void awaitSearchTasksComplete() {
+        TaskManager taskManager = injector.getInstance(TransportService.class).getTaskManager();
+        TimeValue asyncSearchTimeout = SHUTDOWN_SEARCH_TIMEOUT_SETTING.get(this.settings());
+        long millisWaited = 0;
+        while (true) {
+            long searchTasksRemaining = taskManager.getTasks()
+                .values()
+                .stream()
+                .filter(task -> TransportSearchAction.TYPE.name().equals(task.getAction()))
+                .count();
+            if (searchTasksRemaining == 0) {
+                logger.debug("all search tasks complete");
+                return;
+            } else {
+                // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
+                // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
+                // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
+                // be spending on finishing those searches.
+                final TimeValue pollPeriod = TimeValue.timeValueMillis(500);
+                millisWaited += pollPeriod.millis();
+                if (millisWaited >= asyncSearchTimeout.millis()) {
```

Comment:
Since there is now a global timeout, I wonder if we could do without a separate async search timeout. As it is these timeouts need to be aligned (it's useless to have an async search timeout longer than the global timeout).

Reply:
If you don't think we need it, I don't think we need it. At least for now, I've removed it in order to not add surface area. Easy enough to put back should it be necessary. I did plumb the max timeout through to replace the logic here, since if we timeout in this handler we can see how many search tasks were left.